src/libstd/io/buffered.rs

```diff
@@ -451,7 +453,13 @@ impl<W: Write> BufWriter<W> {
     pub fn into_inner(mut self) -> Result<W, IntoInnerError<BufWriter<W>>> {
         match self.flush_buf() {
             Err(e) => Err(IntoInnerError(self, e)),
-            Ok(()) => Ok(self.inner.take().unwrap())
+            Ok(()) => unsafe {
+                let buf = ptr::read(&self.buf);
+                let inner = ptr::read(&self.inner);
+                mem::forget(self);
+                drop(buf);
+                Ok(inner)
+            },
```

Comment:
Why not use a comment here, just "Use ptr::read to extract the inner writer and circumvent our Drop implementation".

I would probably reach for `replace(&mut self.buf, Vec::new())` here before using ptr::read there too, but it doesn't matter much either way I think. The `drop(buf)` is redundant, so that's a bit confusing.